### ISSUE TITLE
There aren't numbers in my Confusion Matrix-YoloV8

### ISSUE THREAD

**Issue by Jes46** [ultralytics/ultralytics#8586]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train, Val

### Bug

Hi all, I've done my train and val mode to object detection and in both modes my confusion matrix looks as you can see below
_________________________________________________________________________________________________
![confusion_matrix_normalized](https://github.com/ultralytics/ultralytics/assets/150959706/d144f25b-700a-411f-afe5-112ec970e06e)
_________________________________________________________________________________________________
It's not possible that it is a labels issue because my model is identifying them:
![val_batch0_pred](https://github.com/ultralytics/ultralytics/assets/150959706/975d18f2-2c91-41c5-a0ad-6390d395f0fa)
![val_batch0_labels](https://github.com/ultralytics/ultralytics/assets/150959706/797a6559-79cb-42f8-a20a-0117b946f995)

Does anyone have any idea why this is happening?

### Environment

Python 3.9.12

### Minimal Reproducible Example

I have done 5 times the train and val mode and the problem happens every time.
In one of them to object classification I get the following matrix:
![confusion_matrix_normalized](https://github.com/ultralytics/ultralytics/assets/150959706/81ba87d6-988a-4a7c-a846-c1499848b2d8)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by Jes46** [comment]
I am using the same images and labels that I used for other trainings and if I was getting the confusion matrix correctly. The bug came up recently when I made a modification to the images in my database where I removed contrast so my Dataset can't be the bug.

The confusion matrix I get is the one automatically provided by YOLO. How do I get the parameters for Y_TRUE and Y_PRED to display it independently?

Finally my model as I show below trains correctly so it would still rule out a labeling problem.

Which libraries would you recommend to upgrade?
![confusion_matrix_normalized](https://github.com/ultralytics/ultralytics/assets/150959706/56aebccf-da11-41b4-845d-24c6a53658e6)

![P_curve](https://github.com/ultralytics/ultralytics/assets/150959706/a8d901ea-8950-472b-bfae-828ff3b162b0)
![PR_curve](https://github.com/ultralytics/ultralytics/assets/150959706/f3a657da-b114-40cc-85f1-e0d35683dff0)
![R_curve](https://github.com/ultralytics/ultralytics/assets/150959706/8246ebc8-f2e5-4aec-aab3-ff56add602ac)
![results](https://github.com/ultralytics/ultralytics/assets/150959706/8b19322f-3c70-4dd3-bd01-9eab069e9eba)
![F1_curve](https://github.com/ultralytics/ultralytics/assets/150959706/6ed7dc1d-347d-47b9-9cdd-5879137522da)

**Comment by Burhan-Q** [comment]
@Jes46 can you share the output from running `yolo checks` in your terminal with your python environment activated?

**Comment by Jes46** [comment]
Yes, of course. Here you have
_____________________________________________________________________________
![yolo checks](https://github.com/ultralytics/ultralytics/assets/150959706/3061cb95-7f9a-413a-b66c-4b489f3c77f9)

**Comment by glenn-jocher** [comment]
@Jes46 thanks for sharing the output from `yolo checks`! 🌟 It looks like your environment is set up correctly. Given that your confusion matrix issue arose after modifying the images in your dataset, it might be worth revisiting those changes, even though the dataset worked fine before. Sometimes, subtle changes in image processing can affect model predictions in unexpected ways.

For the `y_true` and `y_pred` parameters, these are typically extracted during the validation phase. YOLOv8 automatically handles this during training and validation, so you don't need to manually extract these unless you're doing custom evaluations. If you're interested in diving deeper, you could look into the validation code within YOLOv8 to see how predictions and ground truths are being compared.

Regarding library upgrades, it's always a good idea to ensure you're using the latest version of YOLOv8 and its dependencies. You can update YOLOv8 by pulling the latest changes from the GitHub repository and reinstalling any Python dependencies. If you're using a specific branch, make sure it's up to date with the main branch or the latest stable release.

If the issue persists, consider reverting your dataset to its previous state (before the contrast adjustments) to confirm if that's indeed where the problem lies. This could help isolate the cause. Keep us posted! 🚀

**Comment by Burhan-Q** [comment]
@Jes46 it could be due to this line https://github.com/ultralytics/ultralytics/blob/16a91a9b6b307b777ec4e8987aa98c2b6e484ebc/ultralytics/utils/metrics.py#L401
that filters out low values

**Comment by Jes46** [comment]
But What should I do? should I increase the number?

**Comment by Burhan-Q** [comment]
@Jes46 changing the number is likely only going to address the symptom of the numbers not showing, but perhaps not the root cause. First, let's get the confusion matrix values as they are now

```py
import numpy as np
from ultralytics import YOLO

model = YOLO("path/to/best.pt")
results = model.val(data="your_data.yaml")

normalized_matrix = results.confusion_matrix.matrix / (results.confusion_matrix.matrix.sum(0).reshape(1,-1) + 1e-9)

normalized_matrix.nonzero() # check for non-zero values
np.where(normalized_matrix >= 0.005) # check values above threshold
```

The last line here will show how many instances are in your resulting confusion matrix above the `0.005` threshold and the second to last line checks for non-zero values. Both will return a `tuple` with `(row, col)` index locations in `normalized_matrix` where values match the conditions. If you check the count of how many points there are for each, it will be good to know how many are non-zero versus how many are above the threshold.

You can also check both conditions at once

```py
values = np.where((normalized_matrix >= 0.005) & (normalized_matrix > 0)) # checks where both are true

matrix_points = np.array(values).reshape(-1,2)
len(matrix_points ) # how many values are non-zero AND larger than 0.005
```

If there are **no** values both greater than zero AND greater than `0.005` then your issue is more likely with your dataset.

**Comment by pragyahari** [comment]
@Burhan-Q as per your suggested code, getting 163 value, but still there is no value in confusion matrix. I can't figure out the problem. Implementation is performed over yolov10.

**Comment by Khizaralam97** [comment]
> @Burhan-Q as per your suggested code, getting 163 value, but still there is no value in confusion matrix. I can't figure out the problem. Implementation is performed over yolov10.

I had the same issue. I solved it by downgrading matplotlib from 3.8.0 to 3.7.1

**Comment by Burhan-Q** [comment]
@pragyahari make sure you're using the official Ultralytics code and a recent version, presently the latest version is `8.2.63` which should work correctly. If that is still an issue, try downgrading `matplotlib` as suggested by @Khizaralam97 above. This could be a problem with the Notebook/platform and not necessarily with the Ultralytics code.

**Comment by AhmetEkiz** [comment]
> > @Burhan-Q as per your suggested code, getting 163 value, but still there is no value in confusion matrix. I can't figure out the problem. Implementation is performed over yolov10.
> 
> I had the same issue in jupyter notebook. I solved it by downgrading matplotlib from 3.8.0 to 3.7.1

it works for me. Thanks.

`pip install matplotlib==3.7.1`

**Comment by brose649** [comment]
> > @Burhan-Q as per your suggested code, getting 163 value, but still there is no value in confusion matrix. I can't figure out the problem. Implementation is performed over yolov10.
> 
> I had the same issue in jupyter notebook. I solved it by downgrading matplotlib from 3.8.0 to 3.7.1

This worked for me, thank you!
